Tutorials, notebooks, merged in changes from develop (#845)
great-expectations/great_expectations: docs/tutorials/create_expectations.rst

==============================

Creating expectations is an opportunity to blend contextual knowledge from subject-matter experts and insights from
profiling and performing exploratory analysis on your dataset. This tutorial covers creating expectations for a data asset in the Jupyter notebook.

Video
------

If you prefer videos to written tutorials, `James <https://github.com/jcampbell>`_ (one of the original core contributors) walks you through this turorial in a `video on YouTube <https://greatexpectations.io/videos/getting_started/create_expectations>`_.

0. Open Jupyter Notebook
------------------------

This tutorial assumes that:

* you ran ``great_expectations init`` and went through the steps covered in the previous tutorial: :ref:`tutorial_init`.
* your current directory is the root of the project where you ran ``great_expectations init``

You can either follow the tutorial with the dataset that it uses or you can execute the same steps on your project with your own data.

If you get stuck, find a bug or want to ask a question, go to `our Slack <https://greatexpectations.io/slack>`_ - this is the best way to get help from the contributors and other users.

> The dataset is a folder with CSV files containing National Provider Identifier (NPI) data that are processed with pandas.

Reviewer: I might be missing something here but I can't tell how a user would get the data for this tutorial.

Author: The dataset is referenced in the "previous" tutorial - the one that goes over the init command. Let's see if this comes up in user testing sessions.

Jupyter notebooks is the interface for creating expectations.

The ``great_expectations init`` command created ``great_expectations/notebooks/`` folder in your project. The folder contains example notebooks for pandas, Spark and SQL datasources.

If you are following this tutorial using the NPI dataset, open the pandas notebook. If you are working with your dataset, see the instructions for your datasource:

.. content-tabs::

    .. tab-container:: tab0
        :title: pandas

        .. code-block:: bash

            jupyter notebook great_expectations/notebooks/pandas/create_expectations.ipynb

    .. tab-container:: tab1
        :title: pyspark

        .. code-block:: bash

            jupyter notebook great_expectations/notebooks/spark/create_expectations.ipynb

    .. tab-container:: tab2
        :title: SQLAlchemy

        .. code-block:: bash

            jupyter notebook great_expectations/notebooks/sql/create_expectations.ipynb

> 1. Get a DataContext Object

Reviewer: Again, if this were named something less ambiguous like `Project` we could eliminate lots of documentation that basically props up a bad name.

Author: Saving this comment for after 0.8.4

---------------------------

A DataContext represents a Great Expectations project. It organizes datasources, notification settings, data documentation sites, and storage and access for expectation suites and validation results.

> The DataContext is configured via a yml file stored in a directory called great_expectations;

Reviewer: ```suggestion
The DataContext is configured via a yml file stored in a directory called ``great_expectations``.
```

> the configuration file as well as managed expectation suites should be stored in version control.

Reviewer: ```suggestion
This entire directory which includes configuration files as well as expectation suites should be stored in version control.
```

> 

Reviewer: Instantiating a DataContext loads our project configuration and all it's resources.

2. List Data Assets
-------------------

A Data Asset is data you can describe with expectations.

.. content-tabs::

    .. tab-container:: tab0
        :title: pandas

>         A Pandas datasource generates data assets from Pandas DataFrames or CSV files. In this example the pipeline processes NPI data that it reads from CSV files in ``npidata`` directory into Pandas DataFrames. This is the data you want to describe and specify with expectations. That directory and its files are a data asset, named "NPI data" (based on the directory name).

Reviewer: I'm really struggling with the generate language. I wonder if we could make this tutorial much less about teaching about GE and more along the lines of "I just want to make expectations".

Author: Let's see if we can address this in 0.8.5

>         A Pandas datasource generates data assets from Pandas DataFrames or CSV files. In this example the pipeline processes NPI data that it reads from CSV files in ``npidata`` directory into Pandas DataFrames. This is the data you want to describe and specify with expectations. That directory and its files are a data asset, named "NPI data" (based on the directory name).

Reviewer: This sentence feels redundant and doesn't line up w/ `npidata` that is talked about elsewhere:

> That directory and its files are a data asset, named "NPI data" (based on the directory name).

    .. tab-container:: tab1
        :title: pyspark

        A Spark datasource generates data assets from Spark DataFrames or CSV files. In this example the pipeline processes NPI data that it reads from CSV files in ``npidata`` directory into Pandas DataFrames. This is the data you want to describe and specify with expectations. If the example read the data into Spark DataFrames, we would think of this data asset as "data from the npidata directory that we read into Spark DataFrames" and give it a name "NPI data".

    .. tab-container:: tab2
        :title: SQLAlchemy

        A SQLAlchemy datasource generates data assets from tables, views and query results.

        * If the data resided in a table (or view) in a database, it would be accessible as a data asset with the name of that table (or view).
        * If the data did not reside in one table ``npidata`` and, instead, the example pipeline ran an SQL query that fetched the data (probably from multiple tables), the result set of that query would be accessible as a data asset. The name of this data asset would be up to us (e.g., "npidata" or "npidata_query").

> Great Expectations' ``jupyter_ux`` module has a convenience method that lists all data assets and expectation suites known to a Data Context:

Reviewer: ```suggestion
Use this convenience method to list all data assets and expectation suites in your project (using the `DataContext`).
```

.. code-block:: python

    great_expectations.jupyter_ux.list_available_data_asset_names(context)

> Here is the output of this method when executed in our example project:

Reviewer: ```suggestion
The output looks like this:
```

.. image:: ../images/list_data_assets.png
    :width: 600px

``npidata`` is the short name of the data asset. Full names of data assets in a DataContext consist of three parts, for example: ``data__dir/default/npidata``. You don't need to know (yet) how the namespace is managed and the exact meaning of each part. The :ref:`data_context` article describes this in detail.

3. Pick a data asset and set the expectation suite name
-------------------------------------------------------

The ``normalize_data_asset_name`` method converts the short name of a data asset to a full name:

.. code-block:: python

    data_asset_name = "npidata"
    normalized_data_asset_name = context.normalize_data_asset_name(data_asset_name)

>     normalized_data_asset_name

Reviewer: I think we can avoid this discussion entirely since we are not using `normalized_data_asset_name` in this tutorial. I think `get_batch` should do this normalization automatically. Yes it's a little bit of magic, but it would be so much nicer to avoid shoving all our internals out on users.

.. code-block:: python

    expectation_suite_name = "warning"

4. Create a new empty expectation suite
---------------------------------------

> Expectations for a data asset are organized into expectation suites. Usually, we validate data against expectation suites, not single expectations. We recommend 'warning' or 'default' as the name

Reviewer: ```suggestion
Individual Expectations are organized into expectation suites. We recommend 'warning' or 'default' as the name
```

for a first expectation suite associated with a data asset.

> 

Reviewer: Let's create a new empty suite in our project so we can start writing Expectations!

Author: +1

.. code-block:: python

    context.create_expectation_suite(data_asset_name=data_asset_name,
                                     expectation_suite_name=expectation_suite_name)

If an expectation suite with this name already exists for this data_asset, you will get an error. If you would like to overwrite this expectation suite, set ``overwrite_existing=True``.

5. Load a batch of data to create Expectations
----------------------------------------------

Expectations describe data assets. Data assets are composed of batches. Validation checks expectations against a batch of data.

>  For example, a batch could be the most recent day of log data. For a database table, a batch could be the data in that table at a particular time.

Reviewer: remove the tab
```suggestion
For example, a batch could be the most recent day of log data. For a database table, a batch could be the data in that table at a particular time.
```

            .. tab-container:: tab2
                :title: SQLAlchemy

                A SQLAlchemy datasource can accept ``batch_kwargs`` that instruct it load a batch from a table, a view, or a result set of a query:

                If you would like to validate an entire table (or a view) in your database's default schema:

                .. code-block:: python

                    batch_kwargs = {'table': "YOUR TABLE NAME"}

                If you would like to validate an entire table or view from a non-default schema in your database:

                .. code-block:: python

                    batch_kwargs = {'table': "YOUR TABLE NAME", "schema": "YOUR SCHEMA"}

                If you would like to validate using a query to construct a temporary table:

                .. code-block:: python

                    batch_kwargs = {'query': 'SELECT YOUR_ROWS FROM YOUR_TABLE'}

        The examples of ``batch_kwargs`` above can also be the outputs of "generators" used by Great Expectations. You can read about the default Generators' behavior and how to implement additional generators in this article: :ref:`batch_generator`.

> 

Reviewer: remove extra line

|
Now you have the contents of one of the files loaded as batch of the data asset ``data__dir/default/npidata``.

6. Author Expectations
-----------------------

> Now that we have one of the data batches loaded, we can call ``expect`` methods on the data asset in order to check

Reviewer: ```suggestion
Now that we have a batch of data, we can call ``expect`` methods on the data asset in order to check
```

> whether this expectation is true for this batch of data.

Reviewer: This should be ``batch.expect_column_values_to_not_be_null('NPI')``

Author: +1

    df.save_expectation_suite(discard_failed_expectations=False)

8. View the Expectations in Data Docs
-------------------------------------

> Data Docs compiles raw Great Expectations objects including Expectations and Validations into structured documents such as HTML documentation. By default the HTML website is hosted on your local filesystem. When you are working in a team, the website can be hosted in the cloud (e.g., on S3) and serve as the shared source of truth for the team working on the data pipeline.

Reviewer: ```suggestion
Data Docs compiles Expectations and Validations into HTML documentation. By default the HTML website is hosted on your local filesystem. When you are working in a team, the website can be hosted in the cloud (e.g., on S3) and serve as the shared source of truth for the team working on the data pipeline.
```
